cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -82,6 +82,23 @@ void write(Id.Workflow id, RunRecordMeta runRecordMeta, List<ProgramRun> program
     table.put(rowKey, NODES, Bytes.toBytes(value));
   }
 
+  public void delete(Id.Application id) {
+    Scanner scanner = table.scan(null, null);
+    Row row;
+    try {
+      while ((row = scanner.next()) != null) {
+        MDSKey.Splitter splittedKey  = new MDSKey(row.getRow()).split();
+        splittedKey.skipString();
+        String appId = splittedKey.getString();
+        if (appId.equals(id.getId().toString())) {
```

Comment:
this would actually even match applications with a particular name, in all namespaces.